designs/forceful-expiration.md

```diff
@@ -0,0 +1,64 @@
+ 
+**S3:** All Disrupting nodes are tainted as `NoSchedule` to prevent pods from scheduling. Nodes will be marked as Disrupting if necessary capacity for the pods doesn’t initialize within 10 minutes, at which point the taint is removed, and the node goes back to being a Disruption Candidate **(S3 → S2).** All nodes in **S3 through S6** are counted against NDBs, limiting the number of nodes that are nominated **(T2)** for disruption.
+
+**T3 (S3 → S4):** Karpenter waits for necessary capacity in the cluster to be healthy and initialized. Once capacity is available, Karpenter will `kubectl delete` the node(s), tainting the node, beginning graceful termination.
+
+**S4:** Karpenter adds a `NoSchedule` taint and adds an Termination Grace Period Timestamp on the NodeClaim, indicating when Karpenter should transition to Forceful Termination of the node. Karpenter begins draining the node, _respecting PDBs and `do-not-disrupt` pods_. 
+
+***S4** is a point of no return. The node will be terminated within its TGP.* 
+
+**T4 (S4 → S5):** Once the node has no running drainable pods or the TGP timeout is reached, Karpenter directly terminates the underlying instance, initiating forceful termination and cleanup of daemons and networking applications.
+
+**T5 (S5 → S6):** After the termination has been finalized, all network attachments and volumes have been cleaned up, and the node can be removed from the cluster. 
+
+## What’s different in Upstream Karpenter?
+
+Karpenter’s original Expiration semantic was **Forceful**. It was changed to **Graceful** in [this PR](https://github.com/kubernetes-sigs/karpenter/pull/59) so that Karpenter could orchestrate replacement capacity for expired nodes, grouping Expiration with Drift, Underutilization, and Emptiness in **T1**. Users with blocking NDBs, PDBs, or `do-not-disrupt` can prevent expiration from starting, preventing Karpenter from kicking off its TerminationGracePeriod. 
+
+The only difference in the diagram below is Expiration is moved to **T1 (S1 → S2)**.
+
+![](./images/graceful-expiration.png)
+
+## 🔑 Adding Forceful Expiration
+
+### Option 1 - Make Expiration Forceful - Recommended
+
+Revert Expiration’s semantic to be **Forceful**, bypassing past NDBs and `do-not-disrupt`, starting graceful termination as soon as nodes are detected as expired. This will still respect PDBs and `do-not-disrupt` pods while draining, and can be timed out when the user sets a `nodePool.TerminationGracePeriod`.
+
+**Pro:** This natively addresses the cluster administrator's use-case without any user configuration. Making this change is easier to reason about before v1, and harder to do after v1. Adding in Graceful Expiration after v1 is easier, as it's less disruptive. 
+
+**Con:** This removes waiting for pre-spun capacity, increasing the chances that pods aren’t able to reschedule when nodes expire, and is a breaking change.
```

Comment:
I believe you mentioned this during working group but during termination Karpenter would remove the terminating node capacity when calculating how much capacity is needed and possibly start creating nodes?

I think the difference now is there is less time for new node to come up and we don't guarantee the new node is ready before we terminate?

Reply:
One other issue is if you have a broken EC2NodeClass, such as bad AMI or bad config, you will lose capacity as node expires where before it didn't allow this scenario.

Reply:
As soon as the node is terminated, the provisioning controller will trigger and start provisioning nodes for the pods that will be deleted soon. You're right we won't wait to guarantee that the node is ready before we terminate. This is the inherent tradeoff here. Availability (of the pods) vs Compliance.

Reply:
Ah terminated. I was thinking terminat`ing`. Thanks